cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -639,16 +655,31 @@ func (m *McmManager) GetMachineDeploymentNodeTemplate(machinedeployment *Machine
 
 			err := validateNodeTemplate(nodeTemplateAttributes)
 			if err != nil {
-				return nil, fmt.Errorf("NodeTemplate validation error in MachineClass %s : %s", mc.Name, err)
+				return nil, fmt.Errorf("nodeTemplate validation error in MachineClass %s : %s", mc.Name, err)
 			}
 
-			klog.V(1).Infof("Generating node template using nodeTemplate from MachineClass %s", machineClass.Name)
-			instance = instanceType{
-				InstanceType: nodeTemplateAttributes.InstanceType,
-				VCPU:         nodeTemplateAttributes.Capacity["cpu"],
-				Memory:       nodeTemplateAttributes.Capacity["memory"],
-				GPU:          nodeTemplateAttributes.Capacity["gpu"],
+			if len(filteredNodes) > 0 {
+				klog.V(1).Infof("Nodes already existing in the worker pool %s", workerPool)
+				baseNode := filteredNodes[0]
+				klog.V(1).Infof("Worker pool node used to form template is %s and its capacity is cpu: %s, memory:%s", baseNode.Name, baseNode.Status.Capacity.Cpu().String(), baseNode.Status.Capacity.Memory().String())
+				instance = instanceType{
+					VCPU:             baseNode.Status.Capacity[apiv1.ResourceCPU],
+					Memory:           baseNode.Status.Capacity[apiv1.ResourceMemory],
+					GPU:              baseNode.Status.Capacity[gpu.ResourceNvidiaGPU],
+					EphemeralStorage: baseNode.Status.Capacity[apiv1.ResourceEphemeralStorage],
+					NumberOfPods:     baseNode.Status.Capacity[apiv1.ResourcePods],
+				}
+			} else {
+				klog.V(1).Infof("Generating node template only using nodeTemplate from MachineClass %s: template resources-> cpu: %s,memory: %s", machineClass.Name, nodeTemplateAttributes.Capacity.Cpu().String(), nodeTemplateAttributes.Capacity.Memory().String())
+				instance = instanceType{
+					VCPU:   nodeTemplateAttributes.Capacity[apiv1.ResourceCPU],
+					Memory: nodeTemplateAttributes.Capacity[apiv1.ResourceMemory],
+					GPU:    nodeTemplateAttributes.Capacity["gpu"],
+					// Numbers pods per node will depends on the CNI used and the maxPods kubelet config, default is often 110
+					NumberOfPods: resource.MustParse("110"),
+				}
 			}
+			instance.InstanceType = nodeTemplateAttributes.InstanceType
```

Comment:
Just thinking - if we place this assignment here, then there is a potential chance that the `InstanceType` can turn out to be a different one in a worker pool (when the worker pool node is used to form the node template) and thus making the worker pool heterogenous?

Reply:
hmm.. it could happen that a rolling update(due to machineType change for worker pool in shootYaml) is undergoing , and the node which is not yet rolled out is used to form template, in that case old nodeResources will be used, but since the nodeTemplate in machineClass is updated so new `instanceType` is used. This could be a problem.

I'll add a check making sure that the node we pick up has the new instanceType , and if it doesn't then fallback to using the machineClass nodeTemplate only.